pelican/log.py

```diff
@@ -1,142 +1,446 @@
+                }
+            )
+    """
+
+    @classmethod
+    def get_filters(cls, logger):
+        """Return a list of LimitFilters currently on logger."""
+        found = []
+        for x in logger.filters:
+            if isinstance(x, cls):
+                found.append(x)
+        return found
+
+    def __init__(self, threshold=5):
+        """
+        :int threshold:
+            For messages logged with the `limit_msg` in extra, the limit_msg
+            will be used after the threshold is met.
+
+            Example::
+
+                if resource.is_missing:
+                    logger.warning(
+                        'The resource %s is missing', resource.name,
+                        extra={'limit_msg': 'Other resources were missing'}
+                    )
+
+                # if threshold=5, logs:
+                # WARNING: The resource prettiest_cat.jpg is missing
+                # WARNING: The resource best_cat_ever.jpg is missing
+                # WARNING: The resource cutest_cat.jpg is missing
+                # WARNING: The resource lolcat.jpg is missing
+                # WARNING: Other resources were missing
+
+        """
+        super().__init__()
+        self.limit_threshold = None
+        if threshold is not None:
+            # react at, not after, threshold
+            self.limit_threshold = threshold - 1
+        self.group_count = defaultdict(int)
+
+    def filter(self, record):
+        # ``limit_msg`` handling
+        limit_template = record.__dict__.get('limit_msg', None)
+        if limit_template:
+            key = (record.levelno, limit_template)
+            counter = self.group_count[key]
+            if counter > self.limit_threshold:
+                return False  # reject: threshold exceeded
+            elif counter == self.limit_threshold:
+                # change message to the "limit_msg"
+                record.msg = limit_template
+                record.args = record.__dict__.get('limit_args', ())
+
+            self.group_count[key] += 1
+
+        return True  # accept
+
+
+class OnceFilter(logging.Filter):
+    """
+    Allow only the first occurance of matching log messages.
+
+    Similar to the "once" behavior in the ``warnings`` module.
 
-    Groups are specified by the message to use when the number of records in
-    the same group hit the limit.
-    E.g.: log.warning(('43 is not the answer', 'More erroneous answers'))
+    Parameters
+    ----------
+    level: int
+        Only enable log de-duplication for levels equal to or above this
+        value.
     """
 
-    LOGS_DEDUP_MIN_LEVEL = logging.WARNING
+    def __init__(self, level=logging.NOTSET):
+        super().__init__()
+        self.level = level
+        self._ignored = set()
 
-    _ignore = set()
-    _raised_messages = set()
-    _threshold = 5
-    _group_count = defaultdict(int)
+    def ignore(self, level, message):
+        self._ignored.add((level, message))
```

Comment:
While this might have been an artifact of one of the refactors I did in this branch earlier...I think I decided to leave that in as an interface.

Looking at this again now, I'd like to keep it -- I can use it to condense the implementation of ``OnceFilter.filter()`` which is essentially duplicating the same work as part of its process.